### DISCUSSION TITLE
이미지 저장을 S3에 해야하는 이유

### DISCUSSION THREAD

**Discussion by woo-chang** [Busan-Dinosaur/Foodbowl-Backend#108] [기술 도입 문서]
# 개요

현재 이미지 저장은 서버가 동작하고 있는 물리 서버에 저장하도록 되어 있습니다. 서버와 이미지가 같은 위치에 저장되면 어떤 문제가 발생할 수 있었고 서버와 이미지 저장 버킷을 분리했을 때 어떤 장점이 있는지 정리해보면 좋을 것 같습니다 :)

# 참고

- [안전한 S3 사용 가이드](https://techblog.woowahan.com/6217/)

**Comment by woo-chang** [comment]
**장점**

1. 이미지 조회를 위한 요청이 애플리케이션 서버와 동일한 물리 서버로 들어오기에 서버에 영향을 줄 수 있다.
2. 이미지 서버를 우리가 관리하면 이미지 서버를 수평적으로 확장하기 어렵다.
3. S3가 제공해주는 다양한 기능을 활용할 수 있고, 확장에도 용이하다.

**단점**
1. 클라우드 서비스를 사용하기에 비용이 발생한다.
2. S3와 물리적인 서버를 연결하기 위한 추가적인 리소스가 들어간다.

**Comment by woo-chang** [comment]
## 비용

**저장소**

일반적인 저장소로 1TB 저장하는데 한달에 **약 5만원**
- 1TB = 1024GB = 1048576MB
- 사진 하나당 약 2MB이므로 1048576 / 2 = 524288개의 사진이 저장되어야 5만원 정도의 비용발생

서비스 단계에서는 **한달에 만원**도 나오지 않을 것으로 예상

**GET 요청**

조회하는 요청은 천 개당 $0.0004(약 0.5원)의 비용이 발생
- 하루 접속자 1000명 * 조회하는 게시글 100개 * 게시글당 최대 사진 4개 = 400000
- 400000 / 1000 * 0.5 = 200원

매우매우 적은 비용이 발생

**PUT 요청**
저장하는 요청은 천 개당 $0.0055(약 6.5원)의 비용이 발생
- 하루에 작성되는 최대 게시글 50개 * 게시글당 최대 사진 4개 = 200
- 200 / 1000 * 6.5 = 1.3원

매우매우 적은 비용이 발생

---

따라서 S3를 최대한으로 활용해도 사용자가 엄청 많아지지 않는 이상 한달에 만원도 안되는 비용이 발생할 것으로 예상합니다.
그렇기에 위의 장점을 살릴 수 있는 S3 사용을 고려해보면 좋을 것 같습니다 :)

**Comment by Kim0914** [comment]
서버 하나의 컴퓨터에 이미지 / 애플리케이션이 존재하는 것 보다 별도의 스토리지를 사용하는 방법이 더 좋아보이네요 !
S3 도입에 적극 찬성합니다 ㅎㅎ

비용적인 부분도 크게 발생할 것 같지 않아서 도입해도 좋을 것 같아요 😀
